File: l5/z2.pl
halve(X,L,R) :- halve(X,X,L,R).
halve(R,[],[],R) :- !.
halve(R,[_],[],R) :- !.
halve([H|T],[_,_|X],[H|L],R) :- halve(T,X,L,R).

File: l10/z2.hs
halve :: [a] -> ([a], [a])
halve xs = (take half xs, drop half xs)
    where half = (length xs) `div` 2

merge :: Ord a -> ([a], [a]) -> [a]
merge (x, []) = x
merge ([], x) = x
merge ((x:xs), (y:ys)) = case x < y of
    True -> x : merge (xs, (y:ys))
    otherwise -> y : merge ((x:xs), ys)

cross :: (a -> c, b -> d) -> (a, b) -> (c, d)
cross (f, g) = pair (f . fst, g . snd)
pair :: (a -> b, a -> c) -> a -> (b, c)
pair (f, g) x = (f x, g x)

msort :: Ord a => [a] -> [a]
msort [] = []
msort [x] = [x]
msort xs = merge . cross (msort, msort) . halve $ xs
